Add Ecobee humidifier device_info and unique_id

bjpetit opened the pull request:
The humidifier support in the ecobee integration did not
return device_info, causing the humidifier to not be associated
with the ecobee integration or the parent thermostat.
Added unique_id property

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This change adds the device_info and unique_id properties to the ecobee humidifier support code. Without the device_info property the ecobee humidifier is not being linked in as an entity of the ecobee thermostat, and is not associated with the ecobee integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #50046 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bjpetit committed:
Ecobee Integration Humidifier - add device_info and unique_id

The humidifier support in the ecobee integration did not
return device_info, causing the humidifier to not be associated
with the ecobee integration or the parent thermostat.
Also updated humdifier name property
Added unique_id property

bjpetit committed:
Revert humidifier name change

bjpetit commented:
Working through some updates to the humidifier tests...

bieniu commented on homeassistant/components/ecobee/humidifier.py:
```diff
@@ -43,6 +43,40 @@ def __init__(self, data, thermostat_index):
+    @property
+    def unique_id(self):
+        """Return a unique identifier for the humidifier."""
+        return f"{self.data.ecobee.get_thermostat(self.thermostat_index)['identifier']}-humidifier"
```
There is no need to add platform name in `unique_id`. `unique_id`s are platform independent.

bjpetit replied:
I appended the the humidifier string to differentiate the unique_id for the humidifier from that of its parent thermostat. I believe that the humidifier entity and the thermostat need to have separate ids.

On the `self.thermostat` comment. Thanks. I'll simplify that.

MartinHjelmare replied:
As long as the entities are part of different platforms the entity registry will resolve them, even if they have the same unique_id. The entity registry is aware of both the integration domain and the platform domain.

bjpetit replied:
I've removed the extra platform name from the unique_id and simplified the references to the thermostat element.

bieniu commented on homeassistant/components/ecobee/humidifier.py:
```diff
@@ -43,6 +43,40 @@ def __init__(self, data, thermostat_index):
+    @property
+    def device_info(self):
+        """Return device information for the ecobee humidifier."""
+        thermostat = self.data.ecobee.get_thermostat(self.thermostat_index)
```
Why not use `self.thermostat` here?

bjpetit replied:
Updated.

MartinHjelmare commented on homeassistant/components/ecobee/humidifier.py:
```diff
@@ -43,6 +43,40 @@ def __init__(self, data, thermostat_index):
+        try:
+            model = f"{ECOBEE_MODEL_TO_NAME[thermostat['modelNumber']]} Thermostat"
+        except KeyError:
+            _LOGGER.error(
```
We don't allow logging in entity properties. If the log is needed it needs to move to a method eg `async_update`.

bjpetit replied:
Logging removed from this location. I believe that logging in other areas will be sufficient to identify issues.

MartinHjelmare commented:
Adding the unique id shouldn't change the entity_id so there should not be a breaking change.

bjpetit committed:
Merge branch 'dev' of https://github.com/home-assistant/core into ecobee_humid_add_device_info

bjpetit committed:
Add Ecobee humidifier device_info and unique_id

 Address review comments

 Rework humidifier tests to work as entity within
  ecobee device

MartinHjelmare commented on tests/components/ecobee/test_humidifier.py:
```diff
@@ -1,130 +1,79 @@
+@pytest.fixture(name="humidifier")
+def humidifier_fixture(data):
+    """Set up ecobee thermostat object."""
+    return ecobee.EcobeeHumidifier(data, 1)
```
Please don't change the tests like this. They were correct before. We want to use the core interfaces to start the tests and assert core state and call services with the core service registry.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations
